### ISSUE TITLE
Feature request: Button to send to console or to source

### ISSUE THREAD

**Issue by mine-cetinkaya-rundel** [MichelNivard/gptstudio#133]
The RStudio History pane has two buttons: To Console and To Source.

I wonder if these could be implemented instead of or in addition to the Copy button in gptstudio, so instead of copy-paste, one can easily send the code over.

An alternative or additional functionality could be what happens in the RStudio Help pane, where you can highlight a piece of code and Command+Return to send it to the Console.

**Comment by calderonsamuel** [comment]
Thank you for the suggestion! Btw, your OpenIntro IMS1 book is the one I tend to recommend the most on the topic! For now you can achieve this if instead of using the chat app you use the "ChatGPT in source" addin with a text selection. This works in both source and console, but is clearly not that nice to iterate for progressive better output as in the chat app.

For the developers: The To console and To source buttons might be possible if the background job process can talk to `{rstudioapi}` or an environment that can be reached by the global environment. Sadly, I think Command + Return is not possible to implement until RStudio allows to overwrite that shortcut in the viewer pane. Currently it will always send the current line (or selection) in the source document to console, even if you are not focused in it.

**Comment by calderonsamuel** [comment]
An update on this. It is still not possible to send text to console or source when a background job has access to a pointer to the global environment, because a background job is actually a different system process. 

However, I can think of three alternatives:

## Access the R process

 Somehow, be able to run code in the same **process** of the RStudio IDE's R session. The `{processx}` or `{ps}` packages *might* provide some methods for this. I think this could be very difficult as it mostly sounds like "hacking a terminal", which doesn't sound easy or even maintainable.

## Use a lower level Rstudio API

`rstudioapi::insertText()` is a high level wrapper of these lines in the RStudio source code:

https://github.com/rstudio/rstudio/blob/4ce3c93ef1c1b2c277cce08383673b917b37d09e/src/cpp/r/R/Api.R#L1194-L1215

```r
.rs.addApiFunction("selectionSet", function(value = NULL, id = NULL)
{
   # collapse value into single string
   value <- paste(value, collapse = "\n")
   
   # create data payload
   payload <- list(
      value  = .rs.scalar(value),
      doc_id = .rs.scalar(id)
   )
   
   # create request
   request <- .rs.api.createRequest(
      type    = .rs.api.eventTypes$TYPE_SET_EDITOR_SELECTION,
      sync    = TRUE,
      target  = .rs.api.eventTargets$TYPE_ACTIVE_WINDOW,
      payload = payload
   )
   
   # fire away
   .rs.api.sendRequest(request)
})
``` 

Which is itself a wrapper on a [C++ function](https://github.com/rstudio/rstudio/blob/4ce3c93ef1c1b2c277cce08383673b917b37d09e/src/cpp/session/modules/RStudioAPI.cpp#L261)... Which might be another wrapper...

But in the end, we will get to the point where we will find an API that can receive requests (hopefully without authentication), and we could be able to handle that. 

## Escape R

Rstudio is built with Electron.js, and in the end is just a custom browser application. I assume we can take advantage of that and target a) an Electron API, or b) target element id.

Something really funny happens when you open Rstudio -> Help -> Diagnostics -> Toggle Developer Tools, and run the following in the console:

```js
document.getElementsByClassName("ace_cursor")[0].innerText = "test"
```

---

Or ... maybe we should just block the interactive R session? 🤣

**Comment by calderonsamuel** [comment]
I just learnt, completely by accident, of the existence of the [fiery package](https://github.com/thomasp85/fiery), which allows to create "apps" that 1) can run on the main session, 2) can be configured to not block the main session, by a clever use of the {later} package. I think this is it, I'll try to squeeze some time to try it out. 

The biggest problem that comes to my mind is that shiny IS blocking. We would require a new implementation that would:

1) generate the UI
2) run the fiery (or plumber2) server in a non-blocking way
3) link the UI to the server

Potential breaking changes:

1) goodbye shiny
2) goodbye RStudio background processes

So, basically everything we have right now, with exception of the {ellmer} wrappers

Maybe this is interesting for @stevegbrooks

**Comment by stevegbrooks** [comment]
Looking forward to the PR :) thanks
